src/requirements/requirement-graph.ts

```diff
@@ -0,0 +1,91 @@
+import { HashMap, HashSet } from './util/collections';
+
+/**
+ * A mutable graph data structure that represents the mathematically relation between requirement
+ * and course.
+ * An edge between requirement `r` and course `c` implies that `c` may help satisfy `r`. It's up to
+ * the client of the data structure to gradually refine the graph so that an edge `r-c` implies that
+ * `c` definitely helps satisfies `r`.
+ *
+ * We keep type of `Requirement` and `Course` generic, so that they can be easily mocked during
+ * testing and make this basic graph implementation not tied to anything specific representation.
+ */
+export default class RequirementFulfillmentGraph<Requirement, Course> {
+  // Internally, we use a two hash map to represent the bidirection relation
+  // between requirement and courses.
+
+  private readonly requirementToCoursesMap: HashMap<Requirement, HashSet<Course>>;
+
+  private readonly courseToRequirementsMap: HashMap<Course, HashSet<Requirement>>;
+
+  constructor(
+    private readonly getRequirementUniqueID: (r: Requirement) => string,
+    private readonly getCourseUniqueID: (c: Course) => string
+  ) {
+    this.requirementToCoursesMap = new HashMap(getRequirementUniqueID);
+    this.courseToRequirementsMap = new HashMap(getCourseUniqueID);
+  }
+
+  public getAllRequirements(): readonly Requirement[] {
+    return this.requirementToCoursesMap.keys();
+  }
+
+  public getAllCourses(): readonly Course[] {
+    return this.courseToRequirementsMap.keys();
```

Comment:
Are the keys here the course codes? ie CS 2110, INFO 2300, etc..?

Reply:
It might be the course id from course roster so that cross-listed courses are merged into one